solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java

```diff
@@ -118,7 +121,21 @@ private ShardResponse take(boolean bailOnError) {
 
     while (pending.size() > 0) {
       try {
-        Future<ShardResponse> future = completionService.take();
+        SolrRequestInfo reqInfo = SolrRequestInfo.getRequestInfo();
+        Future<ShardResponse> future = null;
+        if(reqInfo != null && reqInfo.getTimeAllowed() > -1 ) {
+          long expireAt = reqInfo.getNOW().getTime() + reqInfo.getTimeAllowed();
+          long currTime = System.currentTimeMillis();
+          if (currTime > expireAt) {
+            throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "Operation timed out");
+          }
+          future = completionService.poll(currTime - expireAt, TimeUnit.MILLISECONDS);
```

Comment:
We have seen threads stuck for more than 10 minutes. Would it be possible to pass timeout in solrclient(solrPeerclient), which requests other shards? That way it will free the resources of the underlying shard request as well. (I am not quite sure about completionService.poll(..) call).

@noblepaul @chatman

Reply:
looks like we configure 10 minute in solr.xml. I feel like it should be at api request level. Any thought on this ?
```
<int name="socketTimeout">${socketTimeout:600000}</int>
```

Reply:
10 minutes is too much anyway